src/thunderbots/software/test/primitive/primitive_factory.cpp

```diff
@@ -0,0 +1,311 @@
+
+    thunderbots_msgs::Primitive prim_msg = pivot_prim.createMsg();
+    std::unique_ptr<Primitive> new_prim =
+        AI::Primitive::createPrimitiveFromROSMessage(prim_msg);
+    std::vector<double> parameters = new_prim->getParameters();
+
+    EXPECT_EQ("Pivot Primitive", new_prim->getPrimitiveName());
+    EXPECT_EQ(robot_id, new_prim->getRobotId());
+    EXPECT_DOUBLE_EQ(pivot_point.x(), parameters[0]);
+    EXPECT_DOUBLE_EQ(pivot_point.y(), parameters[1]);
+    EXPECT_DOUBLE_EQ(final_angle.toRadians(), parameters[2]);
+    EXPECT_DOUBLE_EQ(pivot_radius, parameters[3]);
+    EXPECT_EQ(std::vector<bool>(), new_prim->getExtraBits());
+}
+
+// Test that we can correctly translate a StopPrimitive like:
+// `Primitive` -> `ROS Message` -> `Primitive` and get back the same primitive we
+// started with
+TEST(StopPrimTest, convert_StopPrimitive_to_message_and_back_to_StopPrimitive)
+{
+    const unsigned int robot_id = 2U;
+    const bool coast            = false;
+
+    StopPrimitive stop_prim                  = StopPrimitive(robot_id, coast);
+    thunderbots_msgs::Primitive prim_message = stop_prim.createMsg();
+
+    std::unique_ptr<Primitive> new_prim =
+        AI::Primitive::createPrimitiveFromROSMessage(prim_message);
+
+    // Since we have a `Primitive` and NOT a `StopPrimitive`, we have to check that the
+    // values are correct by getting them from the generic extra_bits array
+    // we use the extra_bits array instead of the parameters array because `StopPrimitive`
+    // has only one variable, coast
+    std::vector<bool> extra_bits = new_prim->getExtraBits();
+
+    EXPECT_EQ("Stop Primitive", new_prim->getPrimitiveName());
+    EXPECT_EQ(robot_id, new_prim->getRobotId());
+    EXPECT_EQ(coast, extra_bits[0]);
+    EXPECT_EQ(new_prim->getParameters(), std::vector<double>());
+    EXPECT_EQ(std::vector<double>(), new_prim->getParameters());
+}
+
+// Test that we can correctly translate a MovePrimitive like:
+// `Primitive` -> `ROS Message` -> `Primitive` and get back the same primitive we
+// started with
+TEST(PrimitiveFactoryTest, convert_DribblePrimitive_to_message_and_back_to_MovePrimitive)
+{
+    const Point destination       = Point(-3, -3);
+    const Angle final_angle       = Angle::ofRadians(4.55);
+    const double final_speed      = 4.22;
+    const double rpm              = 4.5;
+    const bool small_kick_allowed = false;
+    const unsigned int robot_id   = 0U;
+
+    DribblePrimitive dribble_prim = DribblePrimitive(
+        robot_id, destination, final_angle, final_speed, rpm, small_kick_allowed);
+
+    thunderbots_msgs::Primitive prim_message = dribble_prim.createMsg();
+
+    std::unique_ptr<Primitive> new_prim =
+        AI::Primitive::createPrimitiveFromROSMessage(prim_message);
+
+    // Since we have a `Primitive` and NOT a `MovePrimitive`, we have to check that the
+    // values are correct by getting them from the generic parameter array
+    std::vector<double> params  = new_prim->getParameters();
```

Comment:
Just a thought, you can `dynamic_cast` to a `MovePrimitive` and then compare values more explicitly, or make use of equality operators (although we don't define these yet)

Reply:
More of a thought for the future. Equality operators would be useful.

Reply:
Made #377  and #378 to implement the operators in the future. This suggestion isn't really applicable right now (until those are done), so don't worry about it.

Reply:
:+1: can we mark this resolved?